wxSlider layout improvements in wxOSX (#24531 & #24535)

hwiesmann opened the pull request:
Value label is now always centered (#24531) with respect to the slider's moving direction. The labels are adjusted with respect to their position perpendicular to the slider's moving direction in case the slider's across size is smaller than its best size (#24535).

Commit:
BUG FIX #24531 & #24535

Value label is now always centered (#24531) with respect to the slider's moving direction.
The labels are adjusted with respect to their position perpendicular to the slider's moving direction in case the slider's across size is smaller than its best size (#24535).

vadz commented on src/osx/slider_osx.cpp:
```diff
@@ -11,23 +11,64 @@
 
 #if wxUSE_SLIDER
 
+#include <iostream>
```
AFAICS this isn't really needed.

hwiesmann replied:
No, I have forgotten to delete this. It has been used for debugging.

vadz commented on src/osx/slider_osx.cpp:
```diff
@@ -11,23 +11,64 @@
+    // Returns the best dimension of the slider perpendicular to the slider's moving direction
+    int GetBestSliderAcrossDimension(wxSlider const* slider)
+    {
+        return (((slider->GetWindowStyle() & wxSL_AUTOTICKS) != 0) ?
```
This could use more readable `HasFlag()`.

vadz commented on src/osx/slider_osx.cpp:
```diff
@@ -11,23 +11,64 @@
+            if (size.x >= bestSize.x)
+                return GetBestSliderAcrossDimension(slider)+DISTANCE_SLIDER_TEXT;
+            else
+                return MAX(0,size.x-textDimension.x);
```
What is this `MAX`? It results in build errors and I don't see where it is supposed to be defined... We have `wxMax()` in `wx/utils.h` and there is, of course, also `std::max()` and one of them should be used instead.

hwiesmann replied:
Sorry, has been taken from a macOS system header. Will use wxMax instead.

vadz reviewed (changes requested):
Thanks for the fixes but so many things have changed here that it's really difficult to understand what's going on. I guess we should just merge it anyhow to fix a user-visible problem but it would be great to separate changes like this in separate commits if possible.

Also, there are some more concrete problems that have to be fixed before this can be merged, could you please do it? TIA!

hwiesmann commented on src/osx/slider_osx.cpp:
```diff
@@ -11,23 +11,64 @@
-    #define wxSLIDER_DIMENSIONACROSS_WITHTICKMARKS 28
-    #define wxSLIDER_DIMENSIONACROSS_ARROW 21
+    const int wxSLIDER_DIMENSIONACROSS_WITHTICKMARKS = 28;
+    const int wxSLIDER_DIMENSIONACROSS               = 28;
```
I have checked the dimensions with Interface builder (20 for the slider itself plus 2 and 6 for the alignment around it). The old numbers come from an old macOS version, I suppose, as the naming indicates "arrow". But there isn't for a long time anymore an arrow for a slider.
I have also checked the commit. It seems to be that the source code has mainly been copied from Carbon.

hwiesmann commented on src/osx/slider_osx.cpp:
```diff
@@ -346,33 +383,25 @@ wxSize wxSlider::DoGetBestSize() const
-            textwidth = maxtwidth;
-        else
-            textwidth = mintwidth;
+        textheight = MAX(mintheight,maxtheight);
+        textwidth  = MAX(mintwidth, maxtwidth);
```
I would like to do it but the source code had to be basically be re-written. The source code was old inappropriate Carbon code (see my comments below). A lot had to be changed to make it compatible with Cocoa (with respect to alignment and outlook).

PS: There is some more old code inside that can be removed (it is not necessary to change it as it still works but makes the code more complicated than it actually is). I did not remove it because re-writing everything would have been easier. And I already feared that some comments would pop up because of all these changes.

Commit:
BUG FIX #24531 & 24535

Initial commit modified:
 - constexpr instead of const
 - removal of iostream
 - wxMax instead of MAX
 - HasFlag() instead of style & ...

I have still kept all modifications in one commit because the old source code is mostly a copy & paste of old Carbon code that does not make much sense for Cocoa (it worked but that's it). All these modifications are necessary to bring the code from Carbon to Cocoa.
Some further modifications can be done to clean up the code further. But this commit lets actually the sliders look like sliders on Cocoa.

vadz commented:
@csomor If you don't object, I'm going to apply this even if I'm not completely sure about a few things here myself.

csomor commented:
@vadz ok, go ahead, thanks :-)

vadz commented:
I've created #24609 with the same contents as here, so I'm closing this one, please let me know you see anything wrong with the other PR.
